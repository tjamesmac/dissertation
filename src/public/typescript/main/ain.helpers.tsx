import * as React from 'react';
import { IDataPL, IResponse } from './ain.interface';

// State reducer
export const dataReducer = ( state: IDataPL, action: any ) => {
  switch ( action.type ) {
    case 'UPDATE_ORIGINAL':
      return {
        ...state,
        originalString: action.payload,
      };
    case 'UPDATE_NEW':
      return {
        ...state,
        newString: action.payload,
      };
    case 'UPDATE_ORDER':
      const value = action.payload;

      // const mergeStates = state.orderOfWords.concat(value);
      const alternativeState = [...state.orderOfWords, value];

      return {
        ...state,
        orderOfWords: alternativeState,
      };
    case 'UPDATE_DEMO':
      return {
        ...state,
        demographic: action.payload,
      };
      default:
        throw new Error();
  }
};
// removed IResponse from response
export const validateWords = ( response: any, textCheck: string ): string => {
  const keys = Object.keys(response);
  const newKeys = keys.filter( (x) => {
    if (response[x].length) {
      return x;
    } else {
      return null;
    }
  } );
  console.log(newKeys);
  let toChange: string = textCheck;
  for (const key of newKeys) {
    if (textCheck.includes(key)) {
      const word = key;
      const regex = new RegExp(word, 'g');
      // string doesn't replace unless assigned to variable
      const newString = toChange.replace(regex, `<span style='color: green'>${word}</span>`);
      // assign changes to the original string
      toChange = newString;
    } else {
      console.log('oh bother');
    }
  }
  return toChange + '&#8203;';
};

export const createSpan = ( elementTag: string, text: string, colour: string ) => {
  const span = document.createElement(elementTag);
  const textNode = document.createTextNode(text);
  span.appendChild(textNode);
  span.style.color = colour;
  return span;
};
