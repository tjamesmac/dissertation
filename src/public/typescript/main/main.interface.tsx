export interface IWordAndSynonym {
  word: string;
  adjectives?: string[];
  adverbs?: string[];
  nouns?: string[];
  verbs?: string[];
}

export interface IDataPL {
  originalString: string;
  newString: string;
  orderOfWords: string[];
  demographic: string;
}

export interface IResponse {
  [key: string]: string[];
}

export interface IGenderedWords {
  male: string[];
  female: string[];
}
