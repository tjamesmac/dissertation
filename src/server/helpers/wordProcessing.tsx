export interface IWord {
  word: string;
  type: string;
  synonyms: string[];
}

export const isString = (words: string): string[] => {
  const splitString: string[] = words.split(' ');
  return splitString;
};

export const removeDoubles = (splitWords: string[]): string[] => {
  const newArr: string[] = [];
  for ( let i = 0; i < splitWords.length; i++ ) {
    if (!newArr.includes(splitWords[i])) {
      const word: string = splitWords[i].toLowerCase();
      newArr.push(word);
    }
  }
  return newArr;
};

export const getWordData = (splitWords: string[]) => {
  console.log(splitWords);
  const arrayOfAdjectivesMetadata: object[] = [];
  const arrayOfAdjectives: string[] = [];
  const arrayOfOther: string[] = [];

  const fakeWordObjectPositive: any = [
    {
      word: 'hello',
      type: 'adjective',
      synonyms: ['hi', 'hiya'],
    },
    {
      word: 'world',
      type: 'noun',
      synonyms: ['globe', 'the earth'],
    },
    {
      word: 'test',
      type: 'adjective',
      synonyms: ['exam'],
    },
  ];
  // const fakeWordObjectNegative: IWord = {
  //   word: 'world',
  //   type: 'adjective',
  //   synonyms: ['globe', 'the earth'],
  // };
  // // const word: string = splitWords[0];
  // if (!arrayOfAdjectives.includes(word)) {
  //   arrayOfAdjectivesMetadata.push(fakeWordObject);
  //   return arrayOfAdjectivesMetadata;
  // } else {
  //   return null;
  // }

  for (let key of splitWords) {
    const word = key;
    console.log(word, 'top of loop');
    if (!arrayOfAdjectives.includes(word) && !arrayOfOther.includes(word)) {
      // search for the word here


      for (let key of fakeWordObjectPositive) {
        const fakeWord = key.word;
        const fakeType = key.type;
        console.log(fakeWord, 'bottom of loop');
        if (fakeWord === word) {
          console.log(word, fakeWord);
          if (fakeType === 'adjective') {
            arrayOfAdjectivesMetadata.push(key);
            arrayOfAdjectives.push(key.word);
          }
        }
      }
    } else {
      console.log('i am already in the array');
    }
  }
  return arrayOfAdjectivesMetadata;

  // for (let i = 0; i < splitWords.length; i++) {
  //   const word: string = splitWords[i];
  //   if (arrayOfAdjectives.includes(word)) {
  //     // if the adjective is there use the data

  //     arrayOfAdjectivesMetadata.map((item: any) => {
  //       if (item.word === word) {
  //         return item;
  //       }
  //     });
  //     // for (let key in arrayOfAdjectivesMetadata) {
  //     //   if (key.word === word) {
  //     //     return key;
  //     //   }
  //     // }
  //     // here I could potentially just populate the state with the adjective insteadx
  //   } else {
  //     if (!arrayOfAdjectives.includes(word) || !arrayOfOther.includes(word)) {
  //       // inside here I want to send off the request to wordnet
  //       const fakeWordObject: IWord = {
  //         word: 'hello',
  //         type: 'adjective',
  //         synonyms: ['hi', 'hiya'],
  //       };
  //       if (fakeWordObject.type === 'adjective') {
  //         // this needs to push an object into the array
  //         // it will need to include the word, position in the string, list of synonyms
  //         arrayOfAdjectivesMetadata.push(fakeWordObject);
  //         return fakeWordObject;
  //       } else {
  //         arrayOfOther.push(word);
  //       }
  //     }
  //   }
  // }
};
// do I need to record the position of the words so I can then highlight them after wards?
// would be good if I could make a solid loop in order to account for the words
//

// I want to save the submitted text
// I want to save the edited text

// want it to work both ways
//    that is, if you write the words with in the area
//    or submit it via the button
