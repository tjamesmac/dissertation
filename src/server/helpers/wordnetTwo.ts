import WordPos from 'wordpos';

const wordpos = new WordPos();

// right now all i need to do is loop through that array and construct an object with all the possibilities

export async function overAll( inputString: string ) {
  const testString = inputString;

  const getPartOfSpeech = async ( string: string ) => {
    const words = string;

    return await wordpos.getPOS( words, ( response: any ) => {
      return response;
    } );

  };
  // return getPartOfSpeech(testString);
  const confirmPOS = await getPartOfSpeech(testString);
  // console.log(confirmPOS.nouns);
  // console.log(confirmPOS.verbs);
  // console.log(confirmPOS.adjectives);
  // console.log(confirmPOS.adverbs);

  const obj: any = {};
  const returnPOS = (partOfSpeech: string) => {
    switch ( partOfSpeech ) {
      case 'verbs':
        return 'v';
      case 'adjectives':
        return 'a' || 's';
      case 'adverbs':
        return 'r';
      case 'nouns':
        return 'n';
      case 'rest':
        return 'rest';
      default:
        throw new Error();
    }
  };

  const lookupA = async ( array: string[], partOfSpeech: string ) => {
    const wordArray = array;
    const tempObj: any = {};
    console.log(partOfSpeech);
    console.log(returnPOS(partOfSpeech), 'this is the part of speech i need');
    const initialOfPOS = returnPOS(partOfSpeech);
    console.log(initialOfPOS);

    // this was for each but await doesn't work with it
    for (const word of wordArray) {
      tempObj[word] = [];
      tempObj[word][initialOfPOS] = [];
      await wordpos.lookup( word , ( response: any ) => {
        // console.log(word, 'this is called my word');
        console.log(response);
        for ( const element of response ) {
          const synonyms = element.synonyms;
          const speech = element.pos;
          
          console.log(speech, 'this is the speech part that I need to separate objects');
          console.log(synonyms);
          if (initialOfPOS === speech) {
          for ( const w of synonyms ) {
            if ( word !== w ) {
              if (!tempObj[word][initialOfPOS].includes(w)) {

                tempObj[word][initialOfPOS] = [ ...tempObj[word][initialOfPOS], ...[w] ];
              }
            }
          }
        }
        }
      });
    }
    console.log(tempObj);
    return tempObj;
  };
  for ( const pos of Object.keys(confirmPOS) ) {
    
    const partOf = confirmPOS[pos];
    const lookUpResult = await lookupA(await partOf, pos);
    // console.log(lookUpResult);
    obj[pos] = lookUpResult;
    // console.log('onto the next one', pos);
  }
  console.log(obj, 'this is my result object');
  return obj;
  
}
