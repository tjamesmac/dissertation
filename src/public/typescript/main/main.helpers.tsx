
import { IDataPL } from './main.interface';

const genderedWords = {
  maleGenderedWords:
    [
      'active',
      'adventurous',
      'aggress',
      'ambitio',
      'analy',
      'assert',
      'athlet',
      'autonom',
      'boast',
      'challeng',
      'compet',
      'confident',
      'courag',
      'decide',
      'decisive',
      'decision',
      'determin',
      'dominant',
      'domina',
      'force',
      'greedy',
      'headstrong',
      'hierarch',
      'hostil',
      'impulsive',
      'independent',
      'individual',
      'intellect',
      'lead',
      'logic',
      'masculine',
      'objective',
      'opinion',
      'outspoken',
      'persist',
      'principle',
      'reckless',
      'stubborn',
      'superior',
      'self-confiden',
      'self-sufficien',
      'self-relian',
    ]
  ,
  femaleGenderedWords:
    [
      'affectionate',
      'child',
      'cheer',
      'commit',
      'communal',
      'compassion',
      'connect',
      'considerate',
      'cooperat',
      'depend',
      'emotiona',
      'empath',
      'feminine',
      'flatterable',
      'gentle',
      'honest',
      'interpersonal',
      'interdependen',
      'interpersona',
      'kind',
      'kinship',
      'loyal',
      'modesty',
      'nag',
      'nutur',
      'pleasant',
      'polite',
      'quiet',
      'respon',
      'sensitiv',
      'submissive',
      'support',
      'sympath',
      'tender',
      'together',
      'trust',
      'understand',
      'warm',
      'whin',
      'yield',
    ],
};

// State reducer
export const dataReducer = ( state: IDataPL, action: any ) => {
  switch ( action.type ) {
    case 'UPDATE_ORIGINAL':
      return {
        ...state,
        originalString: action.payload,
      };
    case 'UPDATE_NEW':
      return {
        ...state,
        newString: action.payload,
      };
    case 'UPDATE_ORDER':
      const value = action.payload;

      const alternativeState = [...state.orderOfWords, value];

      return {
        ...state,
        orderOfWords: alternativeState,
      };
    case 'UPDATE_DEMO':
      return {
        ...state,
        demographic: action.payload,
      };
    case 'UPDATE_LENGTH':
      return {
        ...state,
        length: action.payload,
      };
      default:
        throw new Error();
  }
};
// removed IResponse from response
// export const validateWords = ( response: any, textCheck: string ): string => {
//   const keys = Object.keys(response);
//   const newKeys = keys.filter( (x) => {
//     if (response[x].length) {
//       return x;
//     } else {
//       return null;
//     }
//   } );
//   let toChange: string = textCheck;
//   for (const key of newKeys) {
//     if (textCheck.includes(key)) {
//       const word = key;
//       // const regex = new RegExp(word, 'g');
//       const regex = new RegExp('\\b' + key + '\\b', 'g');
//       // string doesn't replace unless assigned to variable
//       // styling has been taking up because the replace wasn't capable of handling it
//       const newString = toChange.replace(regex, `<span>${word}</span>`);

//       toChange = newString;
//     } else {
//       console.log('oh bother');
//     }
//   }
//   return toChange + '&#8203;';
// };
const wordMatcher = (key: string, stringToChange: string, gender: string) => {
  
  const word = key;
  const regex = new RegExp('\\b' + key + '\\b', 'g');
  const newString = stringToChange.replace(regex, `<span class='${gender}'>${word}</span>`);

  return newString;
};

export const validateWords = ( response: any, textCheck: string ): any => {
  const maleWords = [];
  const femaleWords = [];
  const keys = Object.keys(response);
  const newKeys = keys.filter( (x) => {
    if (response[x].length) {
      return x;
    } else {
      return null;
    }
  } );
  let toChange: string = textCheck;
  for (const key of newKeys) {
      for (const male of genderedWords.maleGenderedWords) {
        // console.log(key, word);
        // console.log(word.search(key));
        if (textCheck.includes(key) && key.match(male)) {

          const maleNew  = wordMatcher(key, toChange, 'male');
          maleWords.push(key);
          toChange = maleNew;
        }
      }
      for (const female of genderedWords.femaleGenderedWords) {
        if (textCheck.includes(key) && key.match(female)) {
          femaleWords.push(key);
          const femaleWord = wordMatcher(key, toChange, 'female');
          toChange = femaleWord;
        }
      }
  }
  toChange += '&#8203;';
  const returnedObject = { male: maleWords, female: femaleWords, textChange: toChange };
  return returnedObject;
};

export const greenify = async () => {
  const textarea = (document.getElementById('textarea') as HTMLDivElement);
  const children: any = await textarea.children;
  console.log(children);
  if (children) {
    console.log('hello');
    for ( const element of children ) {
      element.style.color = 'green';
    }
  } else {
    console.error('no adjectives are present');
  }
};

export const createSpan = ( elementTag: string, text: string, colour: string ) => {
  const element = document.createElement(elementTag);
  const textNode = document.createTextNode(text);
  element.appendChild(textNode);
  element.style.color = colour;
  return element;
};
