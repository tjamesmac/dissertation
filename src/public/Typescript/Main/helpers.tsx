import { IWord } from '../../../server/helpers/wordProcessing';

export const submission = async (event: React.FormEvent) => {
  event.preventDefault();

  const textAreaValue: string =
  (document.getElementById('textarea') as HTMLDivElement)
  .innerHTML;

  const bodyText: object = {value: textAreaValue};

  try {
    const URL = 'http://localhost:3000/';
    const data = await fetch(URL, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(bodyText),
    });
    const response = await data;
    if (response.status === 200) {
      const responseJson = await response.json();
      console.log(responseJson, 'response');
      validateWords(responseJson, textAreaValue);
    }
  } catch (error) {
    console.error('uh oh error', error);
  }
};

export const validateWords = (response: any, checkString: string) => {
  let toChange = checkString;
  for (let key of response) {
    console.log(key);
    if (checkString.includes(key.word)) {
      console.log('i am in here');
      const find = key.word;
      const regex = new RegExp(find, 'g');
      // string doesn't replace unless assigned to variable
      const newString = toChange.replace(regex, `<span style='color: green'>${key.word}</span>`);
      // assign changes to the og string
      toChange = newString;
      // now I need to create a dropdown menu and populate the synonyms
    } else {
      console.log('oh bother');
    }
  }
  console.log(toChange);
  (document.getElementById('textarea') as HTMLDivElement)
  .innerHTML = toChange;

};
