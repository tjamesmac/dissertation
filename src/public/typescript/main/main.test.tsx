import * as React from 'react';
import renderer from 'react-test-renderer';
import { Main } from './main';
import { validateWords } from './main.helpers';

describe('Main page component', () => {

  it('it renders correctly', () => {
    const tree = renderer
      .create(<Main />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('testing', () => {
  it('should add color to word if it is an adjective', () => {
    const dummyString = 'this is my testing string';
    const dummyData = [
      {
        word: 'testing',
        type: 'adjective',
        synonyms: ['examining'],
      },
    ];
    expect(validateWords(dummyData, dummyString))
      .toEqual(
        {
          initialGendered: {
            female: [],
            male: [],
          },
          updatedString: 'this is my testing string&#8203;',
          valid: ['word', 'type', 'synonyms'],
        },
      );
  });
});
