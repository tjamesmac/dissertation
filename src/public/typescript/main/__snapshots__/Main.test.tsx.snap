// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page component it renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <label
          className="label"
        >
          Please enter a gender:
          <select
            id="select"
            onChange={[Function]}
          >
            <option
              value=""
            >
              -- Please choose an option --
            </option>
            <option
              value="male"
            >
              Male
            </option>
            <option
              value="female"
            >
              Female
            </option>
          </select>
        </label>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          contentEditable={true}
          id="textarea"
          placeholder="Please enter an advert here..."
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-4"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          Check
        </button>
        <button
          className="btn btn-primary btn-extra"
          disabled={false}
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
